### DISCUSSION TITLE
I don't understand surface parameterization

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7501] [Q&A]
Hello,

I tried the `Discrete_conformal_map_parameterizer_3`. I firstly tried with the mesh of a half torus, and I got *"Error: no one-to-one mapping"*. So I tried with the mesh of a half cube. Then I got a matrix of UV coordinates full of (0, 0) except 4 points:

![uv_hcube](https://github.com/CGAL/cgal/assets/4466543/91a256a1-fc76-4d73-8663-89c02bf1e29e)

What can I do with that? 

That's not the first time I tried the surface parameterizations, and I always got strange results. I thought the UV coordinates could help to map for example a checkeboard texture on the mesh, but I don't see how to do that.

**Comment by MaelRL** [comment]
Could you please provide code & input data?

**Comment by stla** [comment]
Yes. Please wait, I deleted the OFF file and I have to redo it. In the meantime I applied the Loop subdivision to my half cube and here are the UV coordinates:

![uv_cubeLoop](https://github.com/CGAL/cgal/assets/4466543/e83917cd-1e20-42aa-abf2-bd52ac343703)

That looks better. But there's still a lot o (0,0).

**Comment by stla** [comment]
Here is the half cube:

![halfcube](https://github.com/CGAL/cgal/assets/4466543/c2b94b9f-4602-4703-83d0-1846ddaa76bc)

And the code : 

```cpp
#ifndef _HEADER_
#include "cgalMesh.h"
#endif

#include <CGAL/Surface_mesh_parameterization/parameterize.h>
#include <CGAL/Surface_mesh_parameterization/Discrete_conformal_map_parameterizer_3.h>
namespace SMP = CGAL::Surface_mesh_parameterization;
typedef K::Point_2                                       Point2;

// [[Rcpp::export]]
Rcpp::NumericMatrix testparam() {
  const std::string filename = "halfcube.off";
  
  Mesh3 sm;
  if(!CGAL::IO::read_polygon_mesh(filename, sm))
  {
    std::cerr << "Invalid input file." << std::endl;
    return EXIT_FAILURE;
  }
  
  // A halfedge on the border
  halfedge_descriptor bhd = CGAL::Polygon_mesh_processing::longest_border(sm).first;
  
  // The 2D points of the uv parametrisation will be written into this map
  typedef Mesh3::Property_map<vertex_descriptor, Point2>  UV_pmap;
  UV_pmap uv_map = sm.add_property_map<vertex_descriptor, Point2>("v:uv").first;
  
  typedef SMP::Discrete_conformal_map_parameterizer_3<Mesh3> Parameterizer;
  
  SMP::Error_code err = SMP::parameterize(sm, Parameterizer(), bhd, uv_map);
  
  if(err != SMP::OK) {
    std::cerr << "Error: " << SMP::get_error_message(err) << std::endl;
    return EXIT_FAILURE;
  }
  
  const size_t nvertices = sm.number_of_vertices();
  Rcpp::NumericMatrix out(nvertices, 2);
  int i = 0;
  for(Mesh3::Vertex_index v : sm.vertices()) {
    Point2 pt = uv_map[v];
    out(i, 0) = pt.x();
    out(i, 1) = pt.y();
    i++;
  }
  
  return out;
}
```


[halfcube.off.txt](https://github.com/CGAL/cgal/files/11675198/halfcube.off.txt)

**Comment by stla** [comment]
What's this joke? There's a lot of duplicated vertices.

**Comment by stla** [comment]
And here is the half torus, without duplicated vertices, clean. Error: no one-to-one mapping.

![halftorus](https://github.com/CGAL/cgal/assets/4466543/c7e004f2-95e4-4106-b457-1d830cc47d6c)
[halftorus.off.txt](https://github.com/CGAL/cgal/files/11678661/halftorus.off.txt)

**Reply by sloriot** [reply]
Some parameterization expects only one boundary, that might be the reason.

**Reply by stla** [reply]
Ah ok. But it doesn't fail with the other half torus. I'm going to try the colors on the human face.

**Comment by stla** [comment]
This works with this human face

![humfaceClean](https://github.com/CGAL/cgal/assets/4466543/d7c8f10a-7563-46af-a483-e1d2875a772b)


![humfaceClean](https://github.com/CGAL/cgal/assets/4466543/ec037bd5-077d-4cb8-af2a-0c2d4e36e74c)

**Comment by stla** [comment]
Here is another half-torus, splitted in another direction. And it works.

```r
> mesh$isValid()
[1] TRUE
> mesh$isValidFaceGraph()
Halfedge Graph Structure is valid.
Face Graph Structure is valid.
[1] TRUE
> mesh$isValidHalfedgeGraph()
Halfedge Graph Structure is valid.
[1] TRUE
> mesh$isValidPolygonMesh()
Halfedge Graph Structure is valid.
Face Graph Structure is valid.
Polygon Mesh Structure is valid.
[1] TRUE
```
![halftorus2](https://github.com/CGAL/cgal/assets/4466543/9bcdd13a-390c-435e-b660-00e29f4992d1)

![uv_halftorus2](https://github.com/CGAL/cgal/assets/4466543/4f9eef7a-76bc-4559-8499-a093e0fa8675)
[halftorus2.off.txt](https://github.com/CGAL/cgal/files/11678920/halftorus2.off.txt)

**Comment by stla** [comment]
Now, what am I supposed to do? I assigned a black color to each vertex with `u < 0.5 && v < 0.5` or with `u > 0.5 && v > 0.5`. All other vertices are white. Here is what I get:

![halftorus2_colors](https://github.com/CGAL/cgal/assets/4466543/cf135d47-661a-4208-96a0-afa035365688)

**Comment by stla** [comment]
Now I tried with an hemisphere. Only one border, guaranteed. But here is what I get:

![hemisphere_colored](https://github.com/CGAL/cgal/assets/4466543/e8629d60-78d8-4d70-87ab-58698ad2378e)

**Comment by stla** [comment]
OH MY GOD I KNOW: THE ORDER OF THE VERTICES IS PERTURBED WHEN I WRITE THE OFF FILE !!!!!

**Reply by MaelRL** [reply]
I haven't had time to look at your issue, but see the maybe-relevant issue: https://github.com/CGAL/cgal/issues/2994

**Reply by stla** [reply]
Maybe this is an additional issue, but the first one is that writing OFF or PLY file changes the order, even before I do the parameterization. I have to investigate that.

**Reply by MaelRL** [reply]
What do you use to write the file? The IO function of CGAL should preserve the order.

**Reply by stla** [reply]
I use `CGAL::IO::write_OFF`. And my mesh is EPECK.

**Comment by stla** [comment]
Yes, that was the problem !!! 
![hemisphere_colored](https://github.com/CGAL/cgal/assets/4466543/f9cc66bd-12ba-4cd3-8c2c-40db2be47334)

**Comment by stla** [comment]
Here are my first 6 mesh faces:

```r
> head(mesh$getFaces())
     [,1] [,2] [,3]
[1,]   31    2    1
[2,]   32    3    2
[3,]   33    4    3
[4,]   34    5    4
[5,]   35    6    5
[6,]   36    7    6
```

Then I observed that the first 6 vertices in the OFF file are vertices 31, 32, 33, 34, 35, 36.

**Comment by stla** [comment]
I think I know: that's because I do a `copy_face_graph` before writing. The goal is to convert EPECK to EPICK because there's a bug when writing an EPECK mesh to a binary PLY file. And `copy_face_graph` changes the order.

**Comment by stla** [comment]
And by the way, since the upgrade to 5.5.2, I cannot read anymore a PLY/OFF file, this crashes the session. This perfectly worked with 5.4.2.

**Comment by stla** [comment]
Here is the disk of uv-coordinates for the Enneper surface:

![uv_Enneper](https://github.com/CGAL/cgal/assets/4466543/6b6a89b8-c48a-4247-bcb4-877483f9ccb3)

As you can see, I colored it with a checkerboard, by grouping the points in squares `[0, 0.25] x [0, 0.25]` etc. The lines of the checkerboard are not very regular. As a consequence, when I color the corresponding vertices on the 3D mesh, I retrieve these non-regular lines:

![Enneper_colored](https://github.com/CGAL/cgal/assets/4466543/cdd22975-cc26-4910-8350-6e6b92e0a95c)

Is there a way to make the lines regular? Of course the quality increases if I add some vertices, but is there another way?

**Comment by stla** [comment]
Wow, I did an isotropic remeshing (but not with CGAL yet) before the parameterization, top!

![Enneper_colored_isotropicRemeshing](https://github.com/CGAL/cgal/assets/4466543/5cbd113e-7547-4789-b135-985692362eeb)

**Comment by stla** [comment]
Now I did it with CGAL.

![Enneper_colored_polar](https://github.com/CGAL/cgal/assets/4466543/99498db7-5dd3-4019-8125-d5fbbbdd9a5d)
![Enneper_colored_circular](https://github.com/CGAL/cgal/assets/4466543/5047558a-c974-42c2-982b-3a66e7ed5f3b)

But why do we see "radial" lines (especially in the blue)?

**Reply by stla** [reply]
> But why do we see "radial" lines (especially in the blue)?

Ooohh because there's no normals!

**Comment by stla** [comment]
Look at this twisted horn. The UV-circle is fully textured with a checkerboard. But the blue squares appear only at the bottom of the horn. Any idea why?

![uv_twisted-horn](https://github.com/CGAL/cgal/assets/4466543/842c426a-b9bf-4d28-a7e0-94d9a8f88b3d)

![twisted-horn](https://github.com/CGAL/cgal/assets/4466543/c4d75e4f-d3df-4f4b-84a1-dadf6c9d4121)

**Reply by MaelRL** [reply]
It looks like the tip of the horn is made of a single square.

You could try an authalic parameterization (tries to preserve areas) rather than a conformal parameterization (tries to preserve angles), such as ARAP or IAP.

**Reply by stla** [reply]
Ok. I firstly tried DAP (discrete authalic). Here is the result.

![twisted-horn-DAP](https://github.com/CGAL/cgal/assets/4466543/11137cf2-64c3-4fd4-bde5-ca03dd1433e9)

**Reply by stla** [reply]
ARAP has crashed the session.

**Reply by stla** [reply]
Strange, it always crashes. Segfault "memory not mapped".

**Reply by stla** [reply]
ARAP works now (it was slow):

![twisted-horn-ARAP](https://github.com/CGAL/cgal/assets/4466543/6ff304be-2797-4aaa-9800-9e004dbd828f)

Indeed, the angles are clearly not preserved, but the areas are not preserved too. Maybe one has to play with the parameter lambda?

I don't understand why this mesh does not appear smooth while it has normals and more than 500000 faces.

**Reply by MaelRL** [reply]
The lambda gives control between angle and area. As lambda goes to infinity, you get more and more rigid (area preserving), but by default it should already be "quite rigid" (1000). You can also try the iterative authalic parameterizer.

**Reply by stla** [reply]
I don't think so. I read the doc. The lambda controls the balance between the angles and *the shape*, not the area. Maybe I misread but I don't think. 

I tried the IAP but it is too slow, I abandonned. 

Still I don't know why this mesh does not appear smooth but after a slight shape smoothing it is perfect. Maybe it's the isotropic remeshing which breaks the smoothness?... ***aahhh I understand right now:*** the isotropic remeshing adds some vertices on the faces, so it does not change the shape, and the mesh does not appear smooth because before the isotropic remeshing it has not a lot of faces.

**Comment by stla** [comment]
I have some plans for today: some animated textures. A rotating radial checkerboard, and a checkerboard transformed by a [Möbius transformation](https://laustep.github.io/stlahblog/frames/pixijs_MobiusCheckerboard.html). Prepare your eyes.

I still tried ARAP, it has always crashed. Wondering whether it is bugged.

**Comment by stla** [comment]
First, here is a small experiment. 

There is a [conformal map between the square and the disk](https://math.stackexchange.com/questions/3159738/complex-transformation-that-transforms-a-square-into-a-circle/3159827#3159827). So I did a checkerboard on the square and I transported it to the disk with this map. The result is the right figure:

![uv_Enneper_psi](https://github.com/CGAL/cgal/assets/4466543/30ee14fa-37f5-4e1f-bd5e-70eaf63c0b8c)

Then I mapped these two circular checkerboards to the Enneper surface, with the conformal parameterization:

![result](https://github.com/CGAL/cgal/assets/4466543/701f3916-698c-42db-a9da-ed1ddb0e9302)

Do you see any striking difference? Not very conclusive... Time to go to the next experiment.

**Comment by stla** [comment]
Here is the first promised animation.

![Enneper-radialCheckerboard-DCP](https://github.com/CGAL/cgal/assets/4466543/a5fc834b-b5c9-4335-8e9a-28585e0f1255)

**Comment by stla** [comment]
The second promised animation. I take a Möbius transformation of order 2 (i.e. `M^2=Id`). Here is for example the transformation of the radial checkerboard with `M^(1.5)`:

![MobiusCheckerboard](https://github.com/CGAL/cgal/assets/4466543/12834460-5729-40eb-b685-e4d1713ca46a)

Then I use `M^t` for the animation, with `t` varying from 0 to 2:

![Enneper-MobiusCheckerboard-DCP](https://github.com/CGAL/cgal/assets/4466543/275b94e3-2d3e-4c92-89e4-1526664588ab)

**Comment by stla** [comment]
Costa surface. It has several borders but the DCP works.

![Costa-DCP](https://github.com/CGAL/cgal/assets/4466543/12d1a9b4-dd60-4521-98ce-2b465260b9ad)

**Comment by stla** [comment]
I mapped an image! Terrific!!

![Enneper_missPsychoCat](https://github.com/CGAL/cgal/assets/4466543/0ef27fe1-c937-4226-88a2-f8d17aafeb16)

**Comment by stla** [comment]
![horn_missPsychoCat](https://github.com/CGAL/cgal/assets/4466543/d8eb497c-740c-431e-8c15-e8cb38df4b4b)

**Comment by stla** [comment]
Togliatti surface

![Togliatti-DCP](https://github.com/CGAL/cgal/assets/4466543/c494edad-5756-4ca3-9772-d227214e09bc)

**Comment by stla** [comment]
Hmm, something strange. Usually when I do a parameterization with a circular border, the UV space I get is a disk centered at (0.5, 0.5) and having radius 0.5. But now, I've just done an ARAP parameterization, and I get a different disk:

![UV-ARAP](https://github.com/CGAL/cgal/assets/4466543/3b247ce4-ff51-44d5-8686-40fce27a37ea)

Is it normal?

**Reply by stla** [reply]
And I also get this disk with the border parameterizer `SMP::Square_border_uniform_parameterizer_3<Mesh3>`!

**Reply by stla** [reply]
Stupid of me, this is a border-free parameterization! I confounded with the DAP I think.

**Comment by stla** [comment]
The Sudanese Möbius band. Unfortunately, this mesh has a "cut", and it breaks the harmony. In fact this cut separates the back side from the front side.

![sudanesMobiusBand](https://github.com/CGAL/cgal/assets/4466543/222cd755-5c16-4eab-b430-4118b422645d)

**Comment by stla** [comment]
I started to write a vignette for my package, about the parameterizations. A vignette of a R package is a short documentation. There is some documentation inside the package but the vignette allows to write more details and is more convenient. 

If you want to take a look, and if you have some feedback it would be welcome. 

This is a HTML document converted to PDF, this is why there are some breaks.

[Parameterizations of surface meshes.pdf](https://github.com/CGAL/cgal/files/11720928/Parameterizations.of.surface.meshes.pdf)

**Comment by stla** [comment]
I've just tried the orbifold borderless parameterization. I used this mesh, with the three cones shown in red:

![soccerstar](https://github.com/CGAL/cgal/assets/4466543/c7fdcb4a-bc0d-4422-a7de-4d70c304da84)

Then I got this uv-space:

![uv-soccerstar](https://github.com/CGAL/cgal/assets/4466543/44796c15-664c-4465-8be2-1bc1390ec936)

How can I use such a parameterization? The points do not fill a nice shape. Maybe I did something bad?

**Reply by MaelRL** [reply]
As ARAP and some other parameterizers, it's a free border parameterizer. It has the advantage that the parameterization is a canonical domain that tiles the 2D space, and you get a seamless parameterization.

See https://noamaig.github.io/html/projects/orbifold/orbifold_lowres.pdf, e.g. Fig 2.

**Reply by stla** [reply]
Not sure to understand. Can we map a texture with such a strange parameterization? I don't see how to do.

**Reply by MaelRL** [reply]
You can check the parameterization plugin in the demo: https://github.com/CGAL/cgal/blob/master/Polyhedron/demo/Polyhedron/Plugins/Surface_mesh/Parameterization_plugin.cpp and in particular the class `Scene_textured_surface_mesh_item`.

**Reply by stla** [reply]
Thanks, I will try to understand that. In the meantime, I took one half of my star to do an easy DCP :)

![halfSoccerStar](https://github.com/CGAL/cgal/assets/4466543/7343cf5c-b51b-4f72-8427-8e12158d3835)
